### DISCUSSION TITLE
Can't send/fetch get request

### DISCUSSION THREAD

**Discussion by lethargistic** [vercel/next.js#55150] [Help] [answered]
### Summary

I'm new to Vercel and Node.js, Express in general. I'm trying to send some data from back-end to front-end and it works well locally, however when I try it on Vercel it doesn't work, instead I get a `GET https://ai-emojify.vercel.app/testGet 404` warning.

I might be misunderstanding something in core.

## Here's the important part of my code

Here's my index.js

```js

`
require("dotenv").config();
const dotenv = require('dotenv');
dotenv.config({ path: '../process.env' });
const express = require("express");
const app = express();
const ntpClient = require('ntp-client');

app.use(express.static('../public'));

app.get("/testGet", (req, res) => {
    // const APIkey = process.env.OPENAI_API_KEY_LATEST
    const test = "help"
    console.log(test)

    res.json({ test });
});

// ntp script (unimportant)
app.get("/vC4B5cmd5ybVouxbC3qnjldC", (req, res) => {
    ntpClient.getNetworkTime("pool.ntp.org", 123, function(err, date) {
        if(err) {
            console.error(err);
            res.status(500).json({ error: "Failed to fetch date" });
            return;
        }

        res.json({ date });
    });
});
console.log(process.env.OPENAI_API_KEY_LATEST)


app.listen(3030, () => {
    console.log("Server is running on port 3030");
});
`
```

And here's my fetch request in HTML:
```
`
    fetch("/testGet")
        .then(response => response.json())
        .then(data => {
            globalLet = data.test

            console.log(globalLet)

            return globalLet;
        })
`
```

The rest of the website runs normally, and it all works when run locally. The website is at the right deployment and my debug didn't lead to much.

Any help is appreciated.

**Comment by Nadeera3784** [comment]
Hi @maksiksking, 
we need to tell Vercel that this is an API and you need to serve this as a serverless function.

for example : 

```
    const express = require('express')

    const app = express()

    app.get('/', (req, res) => {
        res.send('Express JS on Vercel')
    })

    app.get('/testGet', (req, res) => {
        res.json({'message' : 'hello'});
    })

    const port = process.env.PORT || 8080

    app.listen(port, (err, res) => {
        if (err) {
            console.log(err)
            return res.status(500).send(err.message)
        } else {
            console.log('[INFO] Server Running on port:', port)
        }
    })

```

Update NPM scripts : Add start script in scripts object

```
"scripts": {
    "start": "node index.js"
}
```

Add Vercel config in app 


simply we need to add vercel.json file in root of our project. Paste below code in file

```
{
    "version": 2,
    "builds": [
        {
            "src": "index.js",
            "use": "@now/node"
        }
    ],
    "routes": [
        {
            "src": "/(.*)",
            "dest": "index.js"
        }
    ]
}
```

**Comment by lethargistic** [comment]
Hello @Nadeera3784,
Pretty sure I've done all that already, not to say necessarily properly though.

My index.js seems to do everything necessary already, I tried modifying it with your code though.
I had it set to start with nodemon in package.json already and swapping it to node didn't change anything.

Here are the conents of my vercel.json
```
{
  "version": 2,
  "builds": [
    {
      "src": "index.js",
      "use": "@vercel/node"
    },
    {
      "src": "public/**",
      "use": "@vercel/static"
    }
  ],
  "routes":[
    {
      "src": "/api/(.*)",
      "dest": "index.js"
    },
    {
      "src": "/",
      "dest": "public/index.html"
    },
    {
      "src": "/(.+)",
      "dest": "public/$1"
    }
  ]
}
```

And here's my project structure:
![Screenshot_1](https://github.com/vercel/next.js/assets/58907570/e350e339-73cd-4c89-89c4-85bce6ea5b88)
I probably should've included it in the first place

I'll try running a hollow test project and seeing if I can make something from that for my main project.

**Comment by lethargistic** [comment]
Here's a minimum reproducable example.

In theory when I go on the website on Vercel the header should change to "finalyy" like it does when run locally. But it just doesn't, I get a 404 requrest in console instead. I've been searching everywhere and going insane in the last few days, we also have like 45 hours of college per week (not including homework) because of the war so I'm tired as hell and I wanna sleep so bad.

index.js
```
const express = require('express')

// const index = express()

const app = express();

app.use(express.static('./public'));

let reqTest = "finallyy"
console.log(reqTest)

app.get('/yup', (req, res) => {
  console.log(reqTest)
  res.json(reqTest)
})

const port = process.env.PORT || 8080

app.listen(port, (err, res) => {
  if (err) {
    console.log(err)
    return res.status(500).send(err.message)
  } else {
    console.log('[INFO] Server Running on port:', port)
  }
})
```

vercel.json
```
{
  "version": 2,
  "builds": [
    {
      "src": "api/index.js",
      "use": "@vercel/node"
    },
    {
      "src": "public/**",
      "use": "@vercel/static"
    }
  ],
  "routes":[
    {
      "src": "/api/(.*)",
      "dest": "index.js"
    },
    {
      "src": "/",
      "dest": "public/index.html"
    },
    {
      "src": "/(.+)",
      "dest": "public/$1"
    }
  ]
}
```

scrip.js
```
document.addEventListener("DOMContentLoaded", function () {
    const bread = document.getElementById("hel")

    fetch('/yup')
        .then(aye => aye.json())
        .then(aye => {
            const b = JSON.stringify(aye)
            bread.innerHTML = b
        })
    fetch('/yupee')
        .then(aye => aye.json())
        .then(aye => {
            const b = JSON.stringify(aye)
            bread.innerHTML = b
        })
});
```

Vercel also finds my static files and Node:
![Screenshot_5](https://github.com/vercel/next.js/assets/58907570/96bac2e9-f6f9-4228-8d3a-15b3e5bd566d)

Might it be that my static files are overwriting the express routing? I tried using 
`{
  "rewrites": [{ "source": "/api/(.*)", "destination": "/api" }]
}`
in vercel.json but it just loads the html still doesn't send a request.

**Reply by Nadeera3784** [reply]
Hi @maksiksking,
Turns out that, when we deploy(every time) it Vercel  will assign a random domain for you. current URL will be available as `process.env.VERCEL_URL ` and it is CORS protected. even you enable there is no way to pass this URL value to the HTML.
We can try something like EJS to pass values from server to EJS file but remember,  The Vercel platform is meant for static sites and Serverless Functions, it does not support running an Express server. we need to build the pages in order for vercel to provide the content in a static manner. You can try something like nextjs.

https://vercel.com/guides/npm-run-start-not-working

**Comment by lethargistic** [comment, marked as answer]
It took me a lot of trial and error but I managed to make it work. I wasn't fully understanding how serverless functions work and was essentially trying to run a server.

I do feel like it is stated quite vaguely in documentation so for anyone searching this up:
**Everything in /api is run like normal code by Vercel**, and you **don't have to declare where your static files are anywhere** within it - Vercel does it for you you don't need any scripts in package.json and your vercel.json should look like this (if you're using Express):
``
{
  "rewrites": [{ "source": "/api/(.*)", "destination": "/api" }]
}
``

Next time I'll definitely use Next.js for this but this was my first project hosted anywhere and Express seemed to fullfil my purposes well.

Also @Nadeera3784 = chad, thanks!
